x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/inference/deployment/MlPlatformArchitecturesUtil.java

```diff
@@ -0,0 +1,98 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+package org.elasticsearch.xpack.ml.inference.deployment;
+
+import org.elasticsearch.action.ActionListener;
+import org.elasticsearch.action.admin.cluster.node.info.NodesInfoRequestBuilder;
+import org.elasticsearch.action.admin.cluster.node.info.NodesInfoResponse;
+import org.elasticsearch.client.internal.Client;
+import org.elasticsearch.cluster.node.DiscoveryNodeRole;
+import org.elasticsearch.monitor.os.OsInfo;
+import org.elasticsearch.plugins.Platforms;
+import org.elasticsearch.threadpool.ThreadPool;
+import org.elasticsearch.xpack.ml.MachineLearning;
+
+import java.util.Iterator;
+import java.util.Objects;
+import java.util.Set;
+import java.util.stream.Collectors;
+
+import static org.elasticsearch.core.Strings.format;
+
+public class MlPlatformArchitecturesUtil {
+
+    public static ActionListener<NodesInfoResponse> getArchitecturesSetFromNodesInfoResponseListener(
+        ThreadPool threadPool,
+        ActionListener<Set<String>> architecturesListener
+    ) {
+        return ActionListener.wrap(nodesInfoResponse -> {
+            threadPool.executor(MachineLearning.UTILITY_THREAD_POOL_NAME).execute(() -> {
+                architecturesListener.onResponse(getArchitecturesSetFromNodesInfoResponse(nodesInfoResponse));
+            });
+        }, architecturesListener::onFailure);
+    }
+
+    public static NodesInfoRequestBuilder getNodesInfoBuilderWithMlNodeArchitectureInfo(Client client) {
+        return client.admin().cluster().prepareNodesInfo().clear().setNodesIds("ml:true").setOs(true).setPlugins(true);
+    }
```

Comment:
The only places these two methods are called outside of test code they're called in an idiomatic sequence:

```
getArchitecturesSetFromNodesInfoResponseListener
getNodesInfoBuilderWithMlNodeArchitectureInfo
executeAsyncWithOrigin
```

Therefore, I would recommend making these two methods package-protected, so they're accessible from test code but clearly not for general use.

And then add another method to this class that calls the 3 methods in sequence and calls a passed-in listener at the end.

```suggestion
    static ActionListener<NodesInfoResponse> getArchitecturesSetFromNodesInfoResponseListener(
        ThreadPool threadPool,
        ActionListener<Set<String>> architecturesListener
    ) {
        return ActionListener.wrap(nodesInfoResponse -> {
            threadPool.executor(MachineLearning.UTILITY_THREAD_POOL_NAME).execute(() -> {
                architecturesListener.onResponse(getArchitecturesSetFromNodesInfoResponse(nodesInfoResponse));
            });
        }, architecturesListener::onFailure);
    }

    static NodesInfoRequestBuilder getNodesInfoBuilderWithMlNodeArchitectureInfo(Client client) {
        return client.admin().cluster().prepareNodesInfo().clear().setNodesIds("ml:true").setOs(true).setPlugins(true);
    }
```